### PR TITLE
*: rewrite concat using new expression evaluation architecture

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit rewrites `concat` using new expression evaluation architecture.
PTAL @shenli @hanfei1991

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @shenli @hanfei1991

File: expression/builtin_string.go [hunk 1]
```diff
@@ -202,12 +202,34 @@ type concatFunctionClass struct {
...
+		},
+	}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
+func (c *concatFunctionClass) inferType(args []Expression) *types.FieldType {
+	existsBinStr, tp := false, mysql.TypeVarString
+	for i := 0; i < len(args); i++ {
+		curArgTp := args[i].GetType()
+		tp = types.MergeFieldType(tp, curArgTp.Tp)
```
Comment (hanfei1991) [inline expression/builtin_string.go]: concat(1.1, 1.2) will return a float number ??

**Comment by hanfei1991** [comment]
LGTM

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**